File: app/src/main/java/com/example/mike/mp3player/client/views/fragments/PlaybackSpeedControlsFragment.java
package com.example.mike.mp3player.client.views.fragments;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.views.MediaPlayerActionListener;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.fragment.app.Fragment;

import static com.example.mike.mp3player.commons.Constants.DECREASE_PLAYBACK_SPEED;
import static com.example.mike.mp3player.commons.Constants.INCREASE_PLAYBACK_SPEED;

public class PlaybackSpeedControlsFragment extends Fragment {

    private TextView playbackSpeed;
    private AppCompatImageButton increasePlaybackSpeedButton;
    private AppCompatImageButton decreasePlaybackSpeedButton;
    private MediaPlayerActionListener mediaPlayerActionListener;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_playback_speed_controls, container, true);
    }

    @Override
    public void onViewCreated(View view, Bundle bundle) {
        super.onViewCreated(view, bundle);
        this.decreasePlaybackSpeedButton = view.findViewById(R.id.decreasePlaybackSpeed);
        this.decreasePlaybackSpeedButton.setOnClickListener((View v) -> decreasePlaybackSpeed());
        this.increasePlaybackSpeedButton = view.findViewById(R.id.increasePlaybackSpeed);
        this.increasePlaybackSpeedButton.setOnClickListener((View v) -> increasePlaybackSpeed());
        this.playbackSpeed = view.findViewById(R.id.playbackSpeedValue);
    }

    public void setPlaybackState(PlaybackStateCompat state) {
        float speed = state.getPlaybackSpeed();
        if (speed > 0) {
            updatePlaybackSpeedText(speed);
        }
    }

    private void updatePlaybackSpeedText(float speed) {
        Runnable r = () ->  playbackSpeed.setText(getString(R.string.PLAYBACK_SPEED_VALUE, speed));
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(r);
    }

    public void increasePlaybackSpeed() {
        Bundle extras = new Bundle();
        mediaPlayerActionListener.sendCustomAction(INCREASE_PLAYBACK_SPEED, extras);
    }

    public void decreasePlaybackSpeed() {
        Bundle extras = new Bundle();
        mediaPlayerActionListener.sendCustomAction(DECREASE_PLAYBACK_SPEED, extras);
    }

    public void setMediaPlayerActionListener(MediaPlayerActionListener mediaPlayerActionListener) {
        this.mediaPlayerActionListener = mediaPlayerActionListener;
    }
}

File: app/src/main/java/com/github/goldy1992/mp3player/client/views/fragments/PlaybackSpeedControlsFragment.kt
package com.github.goldy1992.mp3player.client.views.fragments

import android.os.Bundle
import android.support.v4.media.session.PlaybackStateCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.VisibleForTesting
import androidx.appcompat.widget.AppCompatImageButton
import com.github.goldy1992.mp3player.R
import com.github.goldy1992.mp3player.client.MediaControllerAdapter
import com.github.goldy1992.mp3player.client.activities.MediaActivityCompat
import com.github.goldy1992.mp3player.client.callbacks.playback.PlaybackStateListener
import com.github.goldy1992.mp3player.commons.Constants
import javax.inject.Inject

class PlaybackSpeedControlsFragment : AsyncFragment(), PlaybackStateListener {
    private var playbackSpeed: TextView? = null
    private var increasePlaybackSpeedButton: AppCompatImageButton? = null
    private var decreasePlaybackSpeedButton: AppCompatImageButton? = null
    private var mediaControllerAdapter: MediaControllerAdapter? = null
    private var speed = 1.0f
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        initialiseDependencies()
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.fragment_playback_speed_controls, container, true)
    }

    override fun onViewCreated(view: View, bundle: Bundle?) {
        super.onViewCreated(view, bundle)
        decreasePlaybackSpeedButton = view.findViewById(R.id.decreasePlaybackSpeed)
        decreasePlaybackSpeedButton.setOnClickListener(View.OnClickListener { v: View? -> decreasePlaybackSpeed() })
        increasePlaybackSpeedButton = view.findViewById(R.id.increasePlaybackSpeed)
        increasePlaybackSpeedButton.setOnClickListener(View.OnClickListener { v: View? -> increasePlaybackSpeed() })
        playbackSpeed = view.findViewById(R.id.playbackSpeedValue)
        // register listeners
        mediaControllerAdapter!!.registerPlaybackStateListener(this)
        //update GUI
        onPlaybackStateChanged(mediaControllerAdapter!!.playbackStateCompat)
    }

    private fun updatePlaybackSpeedText(speed: Float) {
        val r = Runnable { playbackSpeed!!.text = getString(R.string.PLAYBACK_SPEED_VALUE, speed) }
        mainUpdater.post(r)
    }

    @VisibleForTesting
    fun increasePlaybackSpeed() {
        worker!!.post {
            val extras = Bundle()
            mediaControllerAdapter!!.sendCustomAction(Constants.INCREASE_PLAYBACK_SPEED, extras)
        }
    }

    @VisibleForTesting
    fun decreasePlaybackSpeed() {
        worker!!.post {
            val extras = Bundle()
            mediaControllerAdapter!!.sendCustomAction(Constants.DECREASE_PLAYBACK_SPEED, extras)
        }
    }

    override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
        speed = state.playbackSpeed
        if (speed > 0) {
            updatePlaybackSpeedText(speed)
        }
    }

    fun initialiseDependencies() {
        val component = (activity as MediaActivityCompat?)!!.mediaActivityCompatComponent
        component!!.inject(this)
    }

    @Inject
    fun setMediaControllerAdapter(mediaControllerAdapter: MediaControllerAdapter?) {
        this.mediaControllerAdapter = mediaControllerAdapter
    }
}